[Sim Si Leng] iP

pnutzz-0207 opened the pull request:
## SONA

> "The future depends on what you do today." - Mahatma Gandhi 

Hi! My name is sona, a ~~good~~ SUPER AMAZING task manager. I am:

- text based
- easy to learn

Being a personal assistant that helps to manage your task, I do more than just recording tasks!

- [X] Categorize your task into different types: to-do, deadline, event
- [X] mark/unmark your task
- [X] Lookup for tasks on a specific day
- [X] Search tasks by keyword

Most importantly, I am is totally **free**!

It is simple have me in your computer! Simply...

1. Click [here](https://github.com/pnutzz-0207/ip/releases/tag/A-Jar) to download ⏬
2. double-click it 🖱️
3. add your tasks...and TADA! She will manage your task for you ✨

If you want to be a Java programmer, you can use me to practice too! Here is the main method:
```ruby
public static void main(String[] args) {
        new Duke("src/main/data/duke.txt").run();
}
```

Commit:
Add Gradle support

pnutzz-0207 committed:
Update Duke.java

Level-1

pnutzz-0207 committed:
Update Duke.java

Level-2

pnutzz-0207 committed:
Level-3

pnutzz-0207 committed:
Level-4

Bugs not fixed, no exceptions

pnutzz-0207 committed:
Level-5

handled most errors

pnutzz-0207 committed:
Level-6

delete function, change array to arrayList

pnutzz-0207 committed:
A-TextUiTesting

pnutzz-0207 committed:
TextUiTesting

pnutzz-0207 committed:
branch-Level-7

pnutzz-0207 committed:
branch-Level-8

pnutzz-0207 committed:
Merge branch 'branch-Level-8'

pnutzz-0207 committed:
Level-8

pnutzz-0207 committed:
A-moreOOP

pnutzz-0207 committed:
A-Packages

pnutzz-0207 committed:
Create DukeTest.java

pnutzz-0207 committed:
Revert "A-Packages"

This reverts commit 0b4f4c87261d67f9e68cfc924b665c3f9a7ff697.

pnutzz-0207 committed:
A-Packages

pnutzz-0207 committed:
A-JUnit

pnutzz-0207 committed:
branch-A-JavaDoc

pnutzz-0207 committed:
branch-A-CodingStandard

pnutzz-0207 committed:
branch-Level-9

pnutzz-0207 committed:
Merge pull request #1 from pnutzz-0207/branch-A-JavaDoc

Branch A java doc

pnutzz-0207 committed:
branch-A-JavaDoc

pnutzz-0207 committed:
Merge branch 'branch-A-CodingStandard'

pnutzz-0207 committed:
Merge branch 'master' of https://github.com/pnutzz-0207/ip

pnutzz-0207 committed:
Merge branch 'branch-Level-9'

pnutzz-0207 committed:
change packages, add header comments

pnutzz-0207 committed:
Merge branch 'add-gradle-support' into branch-A-Gradle

pnutzz-0207 committed:
checkstyle

pnutzz-0207 committed:
gui

pnutzz-0207 committed:
fix Level-10 bug

AAlghrairy commented on src/main/java/duke/sonautil/Ui.java:
```diff
@@ -0,0 +1,325 @@
+     *
+     * @return message to user
+     */
+    public static String goodbyeMessage() {
+        return "Bye-bye :')  Hope to see you soon!";
```
Perhaps rewrite the method name to be a verb (e.g. returnGoodbyeMessage)?

AAlghrairy commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,37 @@
+
+    private final Duke duke = new Duke("src/main/data/duke.txt");
+
+    @Override
+    public void start(Stage stage) {
```
Perhaps create a header comment for this method (since it is a public method)?

pnutzz-0207 replied:
Yep! Thanks for the reminder!

AAlghrairy commented on src/main/java/duke/sonautil/Parser.java:
```diff
@@ -0,0 +1,236 @@
+     * @return details of user's command
+     * @throws DukeException If command is invalid / not understood
+     * @throws DateTimeParseException If date/time entered is invalid
+     */
+    public String[] messageProcess(String userMessage) throws DukeException, DateTimeParseException {
```
Perhaps rewrite the method name to be in the form of a verb (e.g. "processMessage")? I noticed this issue in a few other places.

pnutzz-0207 replied:
Thanks for the suggestion! it makes sense :)

AAlghrairy commented on src/main/java/duke/sonautil/Parser.java:
```diff
@@ -0,0 +1,236 @@
+            command[1] = String.valueOf(taskIndex);
+
+        } else if (userMessage.startsWith("schedule") && split[0].equals("schedule")) {
+
+            String dateString = userMessage.substring(8).replaceAll(" ", "");
+
+            //no date given
+            if (dateString.equals("")) {
+                throw new DukeException(Ui.scheduleNoDateErrorMessage());
+            }
+
+            //date is invalid
+            LocalDate date = LocalDate.parse(dateString);
+            command[0] = "schedule";
+            command[1] = date.toString();
+
+        } else if (userMessage.startsWith("find") && split[0].equals("find")) {
+
+            //only "find"
+            if (userMessage.length() == 4) {
+                throw new DukeException(Ui.findNoKeywordError());
+            }
+
+            String keyword = userMessage.substring(5);
+
+            //no keyword given
+            if (keyword.replaceAll(" ", "").equals("")) {
+                throw new DukeException(Ui.findNoKeywordError());
+            }
+
+            command[0] = "find";
+            command[1] = keyword;
+
+        } else {
+            command[0] = "unknown";
+            throw new DukeException(Ui.userUnknownInputMessage(userMessage));
+        }
+
+        return command;
+    }
```
Consider using enums here to make the processing of the user input simpler (e.g. create an Enum for task types and pass the type here for easier processing, rather than repeated calls to "startsWith()". Also, any reason why you did not use a switch statement here rather than repeated using "else if"?

AAlghrairy commented on src/main/java/duke/sonautil/Storage.java:
```diff
@@ -0,0 +1,226 @@
+        }
+    }
+
+    /**
+     * deletes the Duke.task from local file by creating a temp file
+     * and copy every Duke.task except the one to be deleted
+     *
+     * @param taskToDelete the index of Duke.task to delete
+     */
```
The "Duke." before each reference to tasks and other Duke related functions is most likely due to a refactor bug (i.e. from changing the name of a package, or moving a file to a new package). I noticed the same issue elsewhere in the code documentation. Consider deleting each of these modifications by the program in the interest of clarity.

```suggestion
    /**
     * Deletes the task from local file by creating a temp file
     * and copy every task except the one to be deleted
     *
     * @param taskToDelete the index of task to delete
     */

AAlghrairy commented on src/main/java/duke/sonautil/Storage.java:
```diff
@@ -0,0 +1,226 @@
+     *
+     * @param taskToDelete the index of Duke.task to delete
+     */
+    public void deleteFromFile(int taskToDelete) {
+        String oldFileName = "src/main/data/duke.txt";
+        String tempFileName = "src/main/data/tempDuke.txt";
+        File tempFile = new File(tempFileName);
+
+        int indexAccessed = 0;
+
+        try (BufferedReader br = new BufferedReader(new FileReader(oldFileName));
+             BufferedWriter bw = new BufferedWriter(new FileWriter(tempFile))) {
+
+            String line;
+            boolean first = true;
+            while ((line = br.readLine()) != null) {
+                if (indexAccessed != taskToDelete) {
+                    if (first) {
+                        bw.write(line);
+                        first = false;
+                    } else {
+                        bw.write("\n" + line);
+                    }
+                }
+                indexAccessed++;
+            }
+
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+
+        File oldFile = new File(oldFileName);
+        oldFile.delete();
+
+        File newFile = new File(tempFileName);
+        newFile.renameTo(oldFile);
+    }
```
I like your implementation of deleting tasks from the file, rather than rewriting the entire file. 👍

AAlghrairy reviewed (commented):
Looks Good 👍.

Just a few coding standard related issues and a few suggestions.

pnutzz-0207 committed:
Update src/main/java/duke/sonautil/Storage.java

Co-authored-by: AAlghrairy <77106335+AAlghrairy@users.noreply.github.com>

pnutzz-0207 committed:
Implement Assertions

There is no assertions that define assumptions made in the program.

Assertions are needed to ensure that there is no potential bug in the code.

Therefore, assertions are implemented into the following parts:
- In TaskList, assert in executeCommand that the keyword and needed information is not null or empty in the String array for their respective user command
- In TaskList, assert in getTask and removeTask to ensure that the task index is in range
- In TaskList, assert in addTask to ensure that the task is instance of Todo, Deadline or Event
- In Task, assert that the description is not empty or null when initiating
- In Storage, assert in changeMarking to ensure the marking strings are valid

Other than assertions, some unused codes are deleted.

pnutzz-0207 committed:
improve quality of code

Currently, some parts of the code has long methods and deep nesting.

Quality of code needs to be improved to make it have better readability.

Changes made are as follows:
- In Parser, processMessage is divided into multiple short and simple methods
- In Storage, code is changed so that it is not in arrowhead style
- In TaskList, executeCommand is divided into multiple short and simple methods

By doing these, it will be easier for the readers to understand the code.

Other minor changes such as changing method names are made.

pnutzz-0207 committed:
Merge pull request #2 from pnutzz-0207/branch-A-CodeQuality

improve quality of code

pnutzz-0207 committed:
Merge branch 'master' into branch-A-Assertions

pnutzz-0207 committed:
Merge pull request #3 from pnutzz-0207/branch-A-Assertions

Implement Assertions

pnutzz-0207 committed:
extension

pnutzz-0207 committed:
fixing minor problems

pnutzz-0207 committed:
gui edit

pnutzz-0207 committed:
improve GUI

pnutzz-0207 committed:
Set theme jekyll-theme-tactile

pnutzz-0207 committed:
add README.md

pnutzz-0207 committed:
Merge branch 'master' of https://github.com/pnutzz-0207/ip

pnutzz-0207 committed:
minor change on UI

pnutzz-0207 committed:
bug fixes
